stub/src/main/java/io/grpc/stub/ServerCalls.java

```diff
@@ -368,8 +375,28 @@ public void onCompleted() {
           throw Status.CANCELLED.withDescription("call already cancelled").asRuntimeException();
         }
       } else {
-        call.close(Status.OK, new Metadata());
-        completed = true;
+
+        if (call.getMethodDescriptor().getType().serverSendsOneMessage()) {
+          Throwable error = null;
+          if (unaryResponse != null) {
+            try {
+              call.sendMessage(unaryResponse);
+            } catch (Throwable t) {
+              error = t;
+            }
+          } else {
+            error = Status.INTERNAL.withDescription("Response message is null for unary call")
+                .asRuntimeException();
+          }
+          if (error != null) {
+            onError(error);
```

Comment:
This error is sent to the client. We should not send either of the two possible `error` values, as we don't want to send exceptions to clients for the application (it's bad practice that we don't want to encourage) and the other message doesn't make sense to the client.

What we should do instead is slightly unclear, but we can talk about it offline.